### PR TITLE
TSL memory layouts

### PR DESCRIPTION
**Pull request by jorendumoulin**
Episode VI: return of the python tests

message to reviewer:
good luck

Todo:

- [x] merge #53 
- [x] add offsets
- [x] add dynamic shape support

# TSL Memory Layouts

## Rationale

A TSL (tiled-strided-layout) memory layout is an MLIR attribute, designed to be used as the `layout` parameter for a `memref` type. A TSL layout tiles the data and defines a stride for every tile, allowing for flexible memory layouts especially suited for hardware accelerators. This layout adds tiling to the existing `StridedLayoutAttr`. While the `AffineMapLayoutAttr` allows for a tiled layout, the representation is not always clear, and more importantly does not allow for non-contiguity, which may be required to maximally exploit the full bandwidth of the memory.

## Notation

We employ the following notation for TSL attributes: (for a 2D matrix and one level of tiling), where the bounds and strides are ordered from outermost -> innermost

`[bound, bound] -> (stride, stride), [bound, bound] -> (stride, stride)`

Consider the following memory layout:
The image represents an `8x8` matrix, where every digit represents the memory address where the element will be stored.

``

In both dimensions, the data is tiled in 2 tiles of size 4, this information is represented with the tiling bounds:

`[2, 4] -> (stride, stride), [2, 4] -> (stride, stride)`

For the first dimension there is a stride within the tile of 4 and across tiles of 32:

`[2, 4] -> (32, 4), [4, 2] -> (stride, stride)`

For the second dimension there is a stride within the tile of 1 and across tiles of 16:

`[2, 4] -> (32, 4), [2, 4] -> (16, 1)`

Additionally, the full TSL layout attribute can also include a base memory offset:

`#tsl.tsl<[2, 4] -> (32, 4), [2, 4] -> (16, 1), offset: 5>`

When no offset is defined, it is assumed to be 0

## Dynamic Sizes

The layout provided allows for some flexibility in defining dynamic shapes within a matrix:

`#tsl.tsl<[?, 4] -> (32, 4), [?, 4] -> (?, 1)>`

The key point is that only the outermost tile is allowed to have dynamic sizes; the sizes and strides of the inner tiles must remain fixed. In the example, the fixed tile sizes are set to 4x4, with strides of 4 and 1. Additionally, there's one extra stride of 32, causing the tiles to be spaced at intervals of 32. The determination of the other strides, once the full matrix dimensions are known, is not yet determined. However, a likely approach is to densely determine the strides from left to right.

For example, if dealing with a 64x64 matrix, the layout would be adjusted accordingly:

`#tsl.tsl<[16, 4] -> (32, 4), [16, 4] -> (?, 1)>`

Here, the missing stride is calculated as 32x16=512. This adjustment ensures that the dynamic shapes remain consistent with the fixed tile sizes and strides while accommodating the overall matrix dimensions.

### CODE REVIEW

File: compiler/ir/tsl/tiled_stride.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    def __init__(self, strides: Sequence[Stride]):
+        self.strides = list(strides)
+
+    def __str__(self) -> str:
+        strides = ", ".join(str(stride.stride) for stride in self.strides)
+        bounds = ", ".join(str(stride.bound) for stride in self.strides)
+        return f"[{strides}] * [{bounds}]"
+
+    def __iter__(self) -> Iterator[list[int, Stride]]:
+        """Returns an iterator over the depths and strides of the Tiled Stride
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_stride.py:37]: unclear

File: compiler/ir/tsl/tiled_stride.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    def depth(self) -> int:
+        """Get the number of strides in the Tiled Stride
+
+        Returns:
+            int: The number of strides in the Tiled Stride.
+        """
+        return len(self.strides)
+
+    def all_values(self) -> Sequence[Sequence[int]]:
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_stride.py:53]: make typing the same as for Stride class

File: compiler/ir/tsl/tiled_stride.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        """Get the stride at a particular depth of the Tiled Stride
+
+        Args:
+            depth (int): The depth of the stride to retrieve.
+
+        Returns:
+            Stride | None: The stride at the specified depth,
+            or None if the depth is invalid.
+        """
+        if depth < 0 or depth >= len(self.strides):
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_stride.py:67]: maybe replace with try catch?

File: compiler/ir/tsl/tiled_strided_layout.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    tstrides: list[TiledStride]
+
+    def __init__(self, tstrides: list[TiledStride]):
+        self.tstrides = tstrides
+
+    def __str__(self) -> str:
+        return "(" + ", ".join(map(str, self.tstrides)) + ")"
+
+    def __iter__(self) -> Iterator[tuple[int, int, Stride]]:
+        """Returns an iterator over the dimensions, depths and
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_strided_layout.py:30]: just over dimension & depth

File: compiler/ir/tsl/tiled_strided_layout.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import numpy as np
+
+from compiler.ir.tsl.stride import Stride
+from compiler.ir.tsl.tiled_stride import TiledStride
+
+
+@dataclass
+class TiledStridedLayout:
+    """TiledStridedLayout is a collection of TiledStrides to represent a tiled
+    strided layout for a multi-dimensional array.
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_strided_layout.py:15]: add comments for dims and depth

File: compiler/ir/tsl/tiled_strided_layout.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        result = result.flatten()
+        return result
+
+    def self_overlaps(self) -> bool:
+        """Check if the Tiled Strided Layout contains overlapping elements"""
+        (
+            u,
+            c,
+        ) = np.unique(self.all_values(), return_counts=True)
+        dup = u[c > 1]  # get duplicates
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_strided_layout.py:74]: variable renaming

File: compiler/ir/tsl/tiled_strided_layout.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        # does not work on illegal workloads
+        if self.self_overlaps():
+            return result
+        if other.self_overlaps():
+            return result
+
+        # find largest contiguous block, starting with stride = 1
+        current_stride = 1
+        while True:
+            # search for contiguous block
```
Comment (jorendumoulin) [inline compiler/ir/tsl/tiled_strided_layout.py:100]: increase contiguous block size

File: compiler/transforms/set_memory_space.py [hunk 8]
```diff
@@ -155,7 +155,7 @@ def match_and_rewrite(self, op: memref.MemorySpaceCast, rewriter: PatternRewrite
             return
 
         # create memref.dim operations for dynamic dimensions
-        shapes = [x.value.data for x in op.results[0].type.shape.data]
+        shapes = [x.data for x in op.results[0].type.shape.data]
```
Comment (jorendumoulin) [inline compiler/transforms/set_memory_space.py:158]: move to other PR

File: tests/filecheck/dialects/tsl.mlir [hunk 9]
```diff
@@ -0,0 +1,7 @@
+// RUN: XDSL_ROUNDTRIP
+
+//CHECK: module
+"builtin.module"() ({
+  //CHECK: #tsl.tsl<([2, 4] * [4, 2], [16, 32] * [4, 2])>
+  %0 = "memref.alloc"() <{"operandSegmentSizes" = array<i32: 0, 0>}> : () -> memref<64x64xindex, #tsl.tsl<([2, 4] * [4, 2], [16, 32] * [4, 2])>, 2 : i32>
+}) : () -> ()
```
Comment (jorendumoulin) [inline tests/filecheck/dialects/tsl.mlir:7]: add newline

**Comment by jorendumoulin** [comment]
Still needs offsets!

File: compiler/ir/tsl/README.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
+# TSL Memory Layouts
+
+## Rationale
+
+A TSL (tiled-strided-layout) memory layout is an MLIR attribute, designed to be used as the `layout` paramter for a `memref` type. A TSL layout tiles the data and defines a stride for every tile, allowing for flexible memory layouts especially suited for hardware accelerators. This layout adds tiling to the existing `StridedLayoutAttr`. While the `AffineMapLayoutAttr` allows for a tiled layout, the representation is not always clear, and more importantly does not allow for non-contiguity, which may be required to maximally exploit the full bandwidth of the memory.
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/README.md:5]:
```suggestion
A TSL (tiled-strided-layout) memory layout is an MLIR attribute, designed to be used as the `layout` parameter for a `memref` type. A TSL layout tiles the data and defines a stride for every tile, allowing for flexible memory layouts especially suited for hardware accelerators. This layout adds tiling to the existing `StridedLayoutAttr`. While the `AffineMapLayoutAttr` allows for a tiled layout, the representation is not always clear, and more importantly does not allow for non-contiguity, which may be required to maximally exploit the full bandwidth of the memory.
```

File: compiler/ir/tsl/README.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+`#tsl.tsl<([4, 32] * [4, 2], [1, 16] * [4, 2], offset: 5)>`
+
+## Dynamic Sizes
+
+The layout provided allows for some flexibility in defining dynamic shapes within a matrix:
+
+`#tsl.tsl<([4, 32] * [4, ?], [1, ?] * [4, ?], offset: 5)>`
+
+The key point is that only the outermost tile is should have dynamic shapes; the sizes and strides of the inner tiles must remain fixed. In the example, the fixed tile sizes are set to 4x4, with strides of 4 and 1. Additionally, there's one extra stride of 32, causing the tiles to be spaced at intervals of 32. The determination of the other strides, once the full matrix dimensions are known, is not yet determined. However, a likely approach is to densely determine the strides from left to right.
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/README.md:40]:
```suggestion
The key point is that only the outermost tile is allowed to have dynamic tile sizes; the sizes and strides of the inner tiles must remain fixed. In the example, the fixed tile sizes are set to 4x4, with strides of 4 and 1. Additionally, there's one extra stride of 32, causing the tiles to be spaced at intervals of 32. The determination of the other strides, once the full matrix dimensions are known, is not yet determined. However, a likely approach is to densely determine the strides from left to right.
```

Comment (JosseVanDelm) [reply compiler/ir/tsl/README.md:40]: Is this a dynamic bound or a dynamic stride?

File: compiler/ir/tsl/README.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+`([stride , stride] x [4, 2], [stride, stride] x [4, 2])`
+
+For the first dimension there is a stride within the tile of 4 and across tiles of 32:
+
+`([4, 32] x [4, 2], [stride, stride] x [4, 2])`
+
+For the second dimension there is a stride within the tile of 1 and across tiles of 16:
+
+`([4, 32] x [4, 2], [1, 16] x [4, 2])`
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/README.md:28]:
```suggestion
`([4, 32] * [4, 2], [1, 16] * [4, 2])`
```

File: compiler/ir/tsl/README.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+<img src="https://github.com/KULeuven-MICAS/snax-mlir/assets/47864363/6d03debe-888e-4e5f-82c2-040434bc1f99 " width="400">
+
+In both dimensions, the data is tiled in 2 tiles of size 4, this information is represented with the tiling bounds:
+
+`([stride , stride] x [4, 2], [stride, stride] x [4, 2])`
+
+For the first dimension there is a stride within the tile of 4 and across tiles of 32:
+
+`([4, 32] x [4, 2], [stride, stride] x [4, 2])`
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/README.md:24]:
```suggestion
`([4, 32] * [4, 2], [stride, stride] * [4, 2])`
```

File: compiler/ir/tsl/README.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+`([stride , stride] x [bound, bound], [stride, stride] x [bound, bound])`
+
+Consider the following memory layout:
+The image represents an `8x8` matrix, where every digit represents the memory address where the element will be stored.
+
+<img src="https://github.com/KULeuven-MICAS/snax-mlir/assets/47864363/6d03debe-888e-4e5f-82c2-040434bc1f99 " width="400">
+
+In both dimensions, the data is tiled in 2 tiles of size 4, this information is represented with the tiling bounds:
+
+`([stride , stride] x [4, 2], [stride, stride] x [4, 2])`
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/README.md:20]:
```suggestion
`([stride , stride] * [4, 2], [stride, stride] * [4, 2])`
```

File: compiler/ir/tsl/README.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+## Rationale
+
+A TSL (tiled-strided-layout) memory layout is an MLIR attribute, designed to be used as the `layout` paramter for a `memref` type. A TSL layout tiles the data and defines a stride for every tile, allowing for flexible memory layouts especially suited for hardware accelerators. This layout adds tiling to the existing `StridedLayoutAttr`. While the `AffineMapLayoutAttr` allows for a tiled layout, the representation is not always clear, and more importantly does not allow for non-contiguity, which may be required to maximally exploit the full bandwidth of the memory.
+
+## Notation
+
+We employ the following notation for TSL attributes: (for a 2D matrix and one level of tiling)
+
+`([stride , stride] x [bound, bound], [stride, stride] x [bound, bound])`
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/README.md:11]:
```suggestion
`([stride , stride] * [bound, bound], [stride, stride] * [bound, bound])`
```

File: compiler/ir/tsl/stride.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+
+@dataclass
+class Stride:
+    """A stride represents the distance between elements in an array.
+    A Stride is defined by a stride and a bound, and the stride can be
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/stride.py:9]: >    A Stride is defined by a stride and a bound ...

This is confusing!

File: tests/ir/tsl/test_tiled_strided_layout.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+
+    ## other way around should be the same
+    lccb2 = tsl2.largest_common_contiguous_block(tsl1)
+    assert len(lccb2) == 2
+    assert lccb2[0].stride == 1
+    assert lccb2[0].bound == 4
+    assert lccb2[1].stride == 4
+    assert lccb2[1].bound == 4
+
+    lccb3 = tsl1.largest_common_contiguous_block(tsl3)
```
Comment (JosseVanDelm) [inline tests/ir/tsl/test_tiled_strided_layout.py:163]: What happens if there's a dynamic tile size in there?

File: tests/ir/tsl/test_tiled_stride.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+def example_strides():
+    return (Stride(1, 4), Stride(4, 6), Stride(24, 2), Stride(None, None))
+
+
+@pytest.fixture()
+def example_tiled_strides(example_strides):
+    stride1, stride2, stride3, dynamic_stride = example_strides
+    tiledStride1 = TiledStride([stride1, stride2])
+    tiledStride2 = TiledStride([stride1, stride2, stride3])
+    tiledStride3 = TiledStride([stride1, dynamic_stride])
```
Comment (JosseVanDelm) [inline tests/ir/tsl/test_tiled_stride.py:17]: I find it a bit confusing that the constructor is written like this:
```
Stride(1, 4), Stride(None, None)
```
Yet the layout is written like:
```
"[1, ?] * [4, ?]"
```

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
I don't fully understand everything I'm afraid, can we discuss offline?

Cool PR!

File: tests/ir/tsl/test_tiled_strided_layout.py (hunk 17 above)
Comment (jorendumoulin) [reply tests/ir/tsl/test_tiled_strided_layout.py:163]: It still works, this case is included in the tests!

File: tests/ir/tsl/test_tiled_strided_layout.py
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply tests/ir/tsl/test_tiled_strided_layout.py:1]: 🤔

File: tests/ir/tsl/test_tiled_strided_layout.py (hunk 17 above)
Comment (jorendumoulin) [reply tests/ir/tsl/test_tiled_strided_layout.py:163]: It just stops searching for a larger contiguous block as soon as it hits a dynamic shape

File: tests/ir/tsl/test_tiled_stride.py (hunk 18 above)
Comment (jorendumoulin) [reply tests/ir/tsl/test_tiled_stride.py:17]: Hmm yes but is this really a problem?

File: compiler/ir/tsl/stride.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    used to generate all values within the bound.
+
+    Args:
+        stride (int | None): The stride of the Stride
+            None represents a dynamic stride
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/stride.py:14]:
```suggestion
            None represents a dynamic step
```

File: compiler/ir/tsl/stride.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    def all_values(self) -> list[int]:
+        """Get all values within the bound of the Stride"""
+        if self.is_dynamic():
+            raise ValueError("Cannot get all values of a dynamic stride")
+        return list(range(0, self.step * self.bound, self.step))
+
+    def __str__(self) -> str:
+        stride = "?" if self.step is None else str(self.step)
+        bound = "?" if self.bound is None else str(self.bound)
+        return f"{stride} x {bound}"
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/stride.py:35]: Is this still valid with the change in format?

Comment (JosseVanDelm) [reply compiler/ir/tsl/stride.py:35]: Would also here change the name to step instead of stride

File: compiler/ir/tsl/tiled_stride.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+@dataclass
+class TiledStride:
+    """
+    A Tiled Stride is a collection of Strides, one for each level of a Tiled Stride.
+
+    The class works with an arbitrary number of tile levels. The most common use case
+    is one level of tiling, which results in a tile depth of 2.
+
+    For example, consider a Tiled Stride for a vector of length 8 of [1, 16] x [4, 2].
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/tiled_stride.py:17]:
```suggestion
    For example, consider a Tiled Stride for a vector of length 8 of [1, 16] -> (4, 2).
```

File: compiler/ir/tsl/tiled_stride.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+    strides: list[Stride]
+
+    def __init__(self, strides: list[Stride]):
+        self.strides = list(strides)
+
+    def __str__(self) -> str:
+        strides = ", ".join(
+            str(stride.step) if stride.step else "?"
+            for stride in reversed(self.strides)
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/tiled_stride.py:34]: Does this mean they are constructed from innermost to outermost? Is this not more confusing?

File: compiler/ir/tsl/tiled_strided_layout.py [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+    def is_dynamic(self) -> bool:
+        """Check if the Tiled Strided Layout is dynamic"""
+        return any(stride.is_dynamic() for _, _, stride in self)
+
+    def dimension(self) -> int:
+        """Get the number of dimensions in the Tiled Strided Layout"""
+        return len(self.tstrides)
+
+    def get_stride(self, dim: int, depth: int) -> Stride:
+        """Get the stride at a particular dimensaion and depth of
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/tiled_strided_layout.py:61]:
```suggestion
        """Get the stride at a particular dimension and depth of
```

File: compiler/parser/tsl_parser.py [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+
+    def _parse_tiled_stride(self) -> TiledStride:
+        """
+        tiled-stride ::= bounds `->` strides
+        """
+        bounds = self._parse_bound()
+        self._parse_token(Token.Kind.ARROW, "Expected arrow")
+        strides = self._parse_stride()
+        if len(strides) != len(bounds):
+            raise ParseError("Expected same number of strides and bounds")
```
Comment (JosseVanDelm) [inline compiler/parser/tsl_parser.py:52]: Please test  these invariants

File: compiler/ir/tsl/tiled_strided_layout.py [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+            return False
+
+        all_values = self.all_values()
+        return np.max(all_values) == len(all_values) - 1
+
+    def largest_common_contiguous_block(
+        self, other: TiledStridedLayout, starting_stride: int = 1
+    ) -> list[Stride]:
+        """
+        Get the largest common contiguous block between two Tiled Strided Layouts.
```
Comment (JosseVanDelm) [inline compiler/ir/tsl/tiled_strided_layout.py:103]:
```suggestion
        Get the largest common contiguous block between two Tiled Strided Layouts.
        Stops searching when it hits a dynamic step, so it finds the largest static size.
```

**Comment by jorendumoulin** [comment]
I resolved all your comments!
Just the negative parsing checks come with some issues:
I included some negative tests, but as you may see, they always check for the following line: `Expected: '>'`, instead of the actual error thrown.

XDSL tries to parse arguments (inclosed in `<attr>`) the following:
```
def in_angle_brackets(self):
        self.parse_punctuation("<")
        try:
            yield
        finally:
            self.parse_punctuation(">")
```

When parsing attr, my own errors will be thrown. This code will then try to just parse `>`, but the attr has not been parsed yet, thus the error looks something like this:

```
blablabla
Error thrown: my own error
blablabla
...
Error thrown: Expected >
```

However, when testing with parsing_diagnostics, only the last line of the error is printed, and I cannot check for my own errors.
```
  %0 = "test.op"() : () -> memref<64x64xindex, #tsl.tsl<[a, b] -> (8, 1), [16, 4] -> (256, 64), offset: 5>, 2 : i32>
                                                         ^
                                                         Expected '>'
```

**Review by JosseVanDelm — APPROVED** [review]

**Comment by JosseVanDelm** [comment]
:tada:

**Pull request merged**